File: cpp/tests/test_person.cpp
#include <gtest/gtest.h>

#include "person.hh"
#include "utils.hh"


namespace tests {

using namespace birthdays;

class PersonTest : public testing::Test {
protected:
    Person p_;
    Person g_;

    PersonTest() : 
        g_{
            "George MiChael", "bluTh", 
            DateArray{1990,3,3}, 
            Person::Relationship::FAMILY
        } {}
};

/// @brief Test formatting of names.
TEST_F(PersonTest, NameFormatting) {
    EXPECT_EQ(g_.first_name(), "george michael");
    EXPECT_EQ(g_.full_name(), "george michael bluth");
}

/// @brief Test relationship to user.
TEST_F(PersonTest, RelationshipTest) {
    EXPECT_EQ(p_.relationship(), Person::Relationship::OTHER);
    EXPECT_EQ(g_.relationship(), Person::Relationship::FAMILY);
}

/// @brief Test date-of-birth.
TEST_F(PersonTest, DobTest) {
    DateArray d1{0,0,0};
    EXPECT_THROW(p_.set_dob(d1), BirthdayError);

    DateArray d2{2025,1,32};
    EXPECT_THROW(p_.set_dob(d2), BirthdayError);

    DateArray d3{2025,2,29};
    EXPECT_THROW(p_.set_dob(d3), BirthdayError);

    DateArray d4{2025,13,1};
    EXPECT_THROW(p_.set_dob(d4), BirthdayError);

    DateArray d5{2024,2,29};  // Leap year.
    EXPECT_NO_THROW(p_.set_dob(d5));
}

/// @brief Test age.
/// @warning Changing *mockDate* changes test results.
TEST_F(PersonTest, AgeTest) {
    DateArray mockDate{2024,12,31};

    EXPECT_EQ(p_.age(), 255);  // Default.
    p_.set_dob(DateArray{2024,1,1});
    p_.set_age_as_of(mockDate);
    EXPECT_EQ(p_.age(), 0);

    g_.set_age_as_of(mockDate);
    EXPECT_EQ(g_.age(), 34);
}

/// @brief Test comparisons.
TEST_F(PersonTest, ComparisonTest) {
    Person gmb{"georGe michaeL", "Bluth", DateArray{1990,3,3}};
    EXPECT_TRUE(g_ == gmb);

    gmb.set_dob(DateArray{1990,3,2});
    EXPECT_TRUE(g_ > gmb);
    gmb.set_dob(DateArray{1990,3,4});
    EXPECT_TRUE(g_ < gmb);
}

}  // End namespace tests.

File: python/tests/test_person.py
import pickle
import tempfile
import unittest
from itertools import product

from pyBirthdays import Person, BirthdayError


class PersonTest(unittest.TestCase):
    def setUp(self):
        self.p = Person()
        self.gmb = Person(
            "George MiChael", "bluTh", 
            [1990,3,3], Person.FAMILY
            )

    def test_default_values(self):
        self.assertEqual(self.p.firstName, "")
        self.assertEqual(self.p.lastName, "")
        self.assertEqual(self.p.dob, [0,0,0])
        self.assertEqual(
            self.p.relationship, 
            Person.OTHER
            )

    def test_types(self):
        self.assertIsInstance(self.gmb.firstName, str)
        self.assertIsInstance(self.gmb.dob, list)
        self.assertIsInstance(self.gmb, Person)

    def test_fullname(self):
        self.assertEqual(
            self.gmb.fullName, "george michael bluth"
        )

    def test_dob(self):
        with self.assertRaises(BirthdayError) as ctx:
            self.p.dob = [0,0,0]
        with self.assertRaises(BirthdayError) as ctx:
            self.p.dob = [2025,1,32]
        with self.assertRaises(BirthdayError) as ctx:
            self.p.dob = [2025,2,29]
        with self.assertRaises(BirthdayError) as ctx:
            self.p.dob = [2024,13,1]

    def test_age(self):
        self.p.dob = [1980,6,7]

        ages = list()
        for m in range(1,13):
            # All days in 2024. Notice discarding invalid dates.
            for today in product([2024],[m],range(1,32)):
                try:
                    self.p.set_age_as_of(list(today))
                except BirthdayError:
                    continue
                ages.append(self.p.age)

        daysBeforeBday = 31 + 29 + 31 + 30 + 31 + 6  # Leap year.
        for a in ages[:daysBeforeBday]:
            self.assertEqual(a, 43)
        for a in ages[daysBeforeBday:]:
            self.assertEqual(a, 44)

    def test_comparison(self):
        p1 = Person("george michael", "bluth", [1990,3,3])
        self.assertEqual(self.gmb, p1)
        p1.lastName = "funke"
        self.assertNotEqual(self.gmb, p1)

    # Test the pickling bindings for Person.
    def test_pickling(self):
        with tempfile.NamedTemporaryFile() as fp:
            pickle.dump(self.gmb, fp)
            fp.flush()  # Ensure data is written to file.
            fp.seek(0)
            loaded_obj = pickle.load(fp)
            self.assertEqual(self.gmb, loaded_obj)
            self.assertEqual(
                self.gmb.relationship, 
                loaded_obj.relationship
            )


if __name__ == "__main__":
    unittest.main()
